File: packages/shared/src/schema/comments.ts
import { z } from 'zod';

import { fileIdSchema, projectIdSchema, userIdSchema } from './ids.js';

export const commentIdSchema = z.string().uuid().brand<'CommentId'>();
export type CommentId = z.infer<typeof commentIdSchema>;

export const commentSchema = z.object({
  id: commentIdSchema,
  projectId: projectIdSchema,
  fileId: fileIdSchema.nullable(),
  parentId: commentIdSchema.nullable(),
  authorId: userIdSchema,
  authorDisplayName: z.string().nullable(),
  // START of the anchored block (1-based line, 0-based column).
  anchorLine: z.number().int().positive().nullable(),
  anchorColumn: z.number().int().nonnegative().nullable(),
  // END of the anchored block. Equal to (anchorLine, anchorColumn)
  // when the comment was made without a selection — degenerates to
  // a point anchor.
  anchorEndLine: z.number().int().positive().nullable(),
  anchorEndColumn: z.number().int().nonnegative().nullable(),
  // The literal text that was selected when the comment was made.
  // Acts as a resilient anchor — when line numbers drift due to
  // later edits, the SPA can re-locate the comment by searching for
  // this string. Capped at 1024 chars server-side.
  anchorSnippet: z.string().nullable(),
  body: z.string(),
  resolvedAt: z.string().nullable(),
  resolvedBy: userIdSchema.nullable(),
  createdAt: z.string(),
  updatedAt: z.string(),
});
export type Comment = z.infer<typeof commentSchema>;

export const createCommentInputSchema = z.object({
  fileId: fileIdSchema.optional(),
  parentId: commentIdSchema.optional(),
  anchorLine: z.number().int().positive().optional(),
  anchorColumn: z.number().int().nonnegative().optional(),
  anchorEndLine: z.number().int().positive().optional(),
  anchorEndColumn: z.number().int().nonnegative().optional(),
  anchorSnippet: z.string().max(1024).optional(),
  body: z.string().trim().min(1, 'Body is required').max(4000),
});
export type CreateCommentInput = z.infer<typeof createCommentInputSchema>;

export const updateCommentInputSchema = z.object({
  body: z.string().trim().min(1).max(4000).optional(),
  resolved: z.boolean().optional(),
});
export type UpdateCommentInput = z.infer<typeof updateCommentInputSchema>;

File: servers/rust/crates/scribe-shared/src/comments.rs
//! Comments + replies. Mirrors `packages/shared/src/schema/comments.ts`.

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

use crate::ids::{CommentId, FileId, ProjectId, UserId};

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Comment {
    pub id: CommentId,
    pub project_id: ProjectId,
    pub file_id: Option<FileId>,
    pub parent_id: Option<CommentId>,
    pub author_id: UserId,
    pub author_display_name: Option<String>,
    /// Start of the anchored block (1-based line, 0-based column).
    pub anchor_line: Option<i32>,
    pub anchor_column: Option<i32>,
    /// End of the anchored block. Equal to (anchor_line, anchor_column)
    /// for point-anchored comments.
    pub anchor_end_line: Option<i32>,
    pub anchor_end_column: Option<i32>,
    /// Literal text that was selected when the comment was made.
    /// Used as a resilient anchor when line numbers drift.
    pub anchor_snippet: Option<String>,
    pub body: String,
    pub resolved_at: Option<DateTime<Utc>>,
    pub resolved_by: Option<UserId>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CreateCommentInput {
    pub file_id: Option<FileId>,
    pub parent_id: Option<CommentId>,
    pub anchor_line: Option<i32>,
    pub anchor_column: Option<i32>,
    pub anchor_end_line: Option<i32>,
    pub anchor_end_column: Option<i32>,
    pub anchor_snippet: Option<String>,
    pub body: String,
}

#[derive(Debug, Clone, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct UpdateCommentInput {
    pub body: Option<String>,
    pub resolved: Option<bool>,
}
